pkg/controller/job/indexed_job_utils_test.go

```diff
@@ -0,0 +1,245 @@
+			{5, v1.PodSucceeded},
+			{6, v1.PodSucceeded},
+			{7, v1.PodSucceeded},
+		},
+		"0-2": {
+			{0, v1.PodSucceeded},
+			{1, v1.PodFailed},
+			{1, v1.PodSucceeded},
+			{2, v1.PodSucceeded},
+			{2, v1.PodSucceeded},
+			{3, v1.PodFailed},
+		},
+		"0,2-5": {
+			{0, v1.PodSucceeded},
+			{1, v1.PodFailed},
+			{2, v1.PodSucceeded},
+			{3, v1.PodSucceeded},
+			{4, v1.PodSucceeded},
+			{5, v1.PodSucceeded},
+		},
+	}
+	for want, tc := range cases {
+		t.Run(want, func(t *testing.T) {
+			pods := hollowPodsWithIndexPhase(tc)
+			gotStr := calculateCompletedIndexesStr(pods)
+			if diff := cmp.Diff(want, gotStr); diff != "" {
+				t.Errorf("Unexpected completed indexes (-want,+got):\n%s", diff)
+			}
+		})
+	}
+}
+
+func TestFirstPendingIndexes(t *testing.T) {
+	cases := map[string]struct {
+		cnt         int
+		completions int
+		pods        []indexPhase
+		want        []int
+	}{
+		"cnt greater than completions": {
+			cnt:         5,
+			completions: 3,
+			want:        []int{0, 1, 2},
+		},
+		"cnt less than completions": {
+			cnt:         2,
+			completions: 5,
+			want:        []int{0, 1},
+		},
+		"first pods running or succeeded": {
+			pods: []indexPhase{
+				{0, v1.PodRunning},
+				{1, v1.PodPending},
+				{2, v1.PodFailed},
+			},
+			cnt:         3,
+			completions: 10,
+			want:        []int{2, 3, 4},
+		},
+		"last pods running or succeeded": {
+			pods: []indexPhase{
+				{4, v1.PodFailed},
+				{5, v1.PodSucceeded},
+				{6, v1.PodPending},
+			},
+			cnt:         6,
+			completions: 6,
+			want:        []int{0, 1, 2, 3, 4},
+		},
+		"mixed": {
+			pods: []indexPhase{
+				{1, v1.PodFailed},
+				{2, v1.PodSucceeded},
+				{3, v1.PodPending},
+				{5, v1.PodFailed},
+				{5, v1.PodRunning},
+				{8, v1.PodPending},
+			},
+			cnt:         5,
+			completions: 10,
+			want:        []int{0, 1, 4, 6, 7},
+		},
+	}
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			pods := hollowPodsWithIndexPhase(tc.pods)
+			got := firstPendingIndexes(pods, tc.cnt, tc.completions)
+			if diff := cmp.Diff(tc.want, got); diff != "" {
+				t.Errorf("Wrong first pending indexes (-want,+got):\n%s", diff)
+			}
+		})
+	}
+}
+
+func TestAppendDuplicatedIndexPodsForRemoval(t *testing.T) {
+	cases := map[string]struct {
+		pods     []indexPhase
+		wantRm   []indexPhase
+		wantLeft []indexPhase
+	}{
+		"all unique": {
+			pods: []indexPhase{
+				{unknownCompletionIndex, v1.PodPending},
+				{2, v1.PodPending},
+				{5, v1.PodRunning},
+			},
+			wantRm: []indexPhase{
+				{unknownCompletionIndex, v1.PodPending},
+			},
+			wantLeft: []indexPhase{
+				{2, v1.PodPending},
+				{5, v1.PodRunning},
+			},
+		},
+		"all with index": {
+			pods: []indexPhase{
+				{5, v1.PodPending},
+				{0, v1.PodRunning},
+				{3, v1.PodPending},
+				{0, v1.PodRunning},
+				{3, v1.PodRunning},
+				{0, v1.PodPending},
+			},
+			wantRm: []indexPhase{
+				{0, v1.PodPending},
+				{0, v1.PodRunning},
+				{3, v1.PodPending},
```

Comment:
In this test case I'm missing information which of the two running 0's got removed, to make sure we don't remove the older one. Namely to ensure that sorting inside `appendPodsWithSameIndexForRemovalAndRemaining` doesn't accidentally disappear.

Reply:
sgtm, I'm still to look at that other PR.